### ISSUE TITLE
Improvement : GitLab upgrades

### ISSUE THREAD

**Issue by cchaudier** [geerlingguy/ansible-role-gitlab#165]
Hello @geerlingguy,  
  
Whe work on a global french instance of GitLab : [http://beta.foggit.fr](http://beta.foggit.fr/)

  
We use this role for instaling GitLab and we would like to use the role to update our GitLab, currently this is not possible.  
  
When we modify the value of the variable `gitlab_version` with a version higher than the one installed, the task [`Install GitLab`](https://github.com/geerlingguy/ansible-role-gitlab/blob/374dcf1d2a57d64ee83b9733af532ba51e45445c/tasks/main.yml#L42) is skipped because there is already a GitLab installed and therefore the file `/usr/bin/gitlab-ctl` exists .  
  
So we tried to relaunch the role by deleting the `/usr/bin/gitlab-ctl` file, the task is no longer skipped but the "install gitlab" task fails with an error.  

Here is the error   `/var/lib/dpkg/tmp.ci/preinst: gitlab-ctl: not found`

```
"cache_update_time": 1600210091,
"cache_updated": false,
"changed": false,
"msg": "'/usr/bin/apt-get -y -o \"Dpkg::Options::=--force-confdef\" -o \"Dpkg::Options::=--force-confold\"      install 'gitlab-ce=13.2.9-ce.0'' failed: E: Sub-process /usr/bin/dpkg returned an error code (1)\n",
"rc": 100,
"stderr": "E: Sub-process /usr/bin/dpkg returned an error code (1)\n",
"stderr_lines": ["E: Sub-process /usr/bin/dpkg returned an error code (1)"],
"stdout": "Reading package lists...\nBuilding dependency tree...\nReading state information...\nThe following packages were automatically installed and are no longer required:\n  libgnutls-dane0 libunbound8\nUse 'sudo apt autoremove' to remove them.\nThe following packages will be upgraded:\n  gitlab-ce\napt-listchanges : Lecture des fichiers de modifications (« changelog »)...\n1 upgraded,
0 newly installed,
0 to remove and 88 not upgraded.\nNeed to get 0 B/769 MB of archives.\nAfter this operation,
30.7 kB of additional disk space will be used.\n(Reading database ... \r(Reading database ... 5%\r(Reading database ... 10%\r(Reading database ... 15%\r(Reading database ... 20%\r(Reading database ... 25%\r(Reading database ... 30%\r(Reading database ... 35%\r(Reading database ... 40%\r(Reading database ... 45%\r(Reading database ... 50%\r(Reading database ... 55%\r(Reading database ... 60%\r(Reading database ... 65%\r(Reading database ... 70%\r(Reading database ... 75%\r(Reading database ... 80%\r(Reading database ... 85%\r(Reading database ... 90%\r(Reading database ... 95%\r(Reading database ... 100%\r(Reading database ... 143504 files and directories currently installed.)\r\nPreparing to unpack .../gitlab-ce_13.2.9-ce.0_amd64.deb ...\r\n/var/lib/dpkg/tmp.ci/preinst: 41: /var/lib/dpkg/tmp.ci/preinst: gitlab-ctl: not found\r\ndpkg: error processing archive /var/cache/apt/archives/gitlab-ce_13.2.9-ce.0_amd64.deb (--unpack):\r\n new gitlab-ce package pre-installation script subprocess returned error exit status 1\r\n/var/lib/dpkg/info/gitlab-ce.postinst: 114: /var/lib/dpkg/info/gitlab-ce.postinst: /opt/gitlab/embedded/bin/symlink_ctl_cmds: not found\r\ncp: cannot stat '/opt/gitlab/etc/gitlab.rb.template': No such file or directory\r\nsed: can't read /etc/gitlab/gitlab.rb: No such file or directory\r\nchmod: cannot access '/etc/gitlab/gitlab.rb': No such file or directory\r\nErrors were encountered while processing:\r\n /var/cache/apt/archives/gitlab-ce_13.2.9-ce.0_amd64.deb\r\nneedrestart is being skipped since dpkg has failed\n",
"stdout_lines": ["Reading package lists...",
"Building dependency tree...",
"Reading state information...",
"The following packages were automatically installed and are no longer required:",
"  libgnutls-dane0 libunbound8",
"Use 'sudo apt autoremove' to remove them.",
"The following packages will be upgraded:",
"  gitlab-ce",
"apt-listchanges : Lecture des fichiers de modifications (« changelog »)...",
"1 upgraded,
0 newly installed,
0 to remove and 88 not upgraded.",
"Need to get 0 B/769 MB of archives.",
"After this operation,
30.7 kB of additional disk space will be used.",
"(Reading database ... ",
"(Reading database ... 5%",
"(Reading database ... 10%",
"(Reading database ... 15%",
"(Reading database ... 20%",
"(Reading database ... 25%",
"(Reading database ... 30%",
"(Reading database ... 35%",
"(Reading database ... 40%",
"(Reading database ... 45%",
"(Reading database ... 50%",
"(Reading database ... 55%",
"(Reading database ... 60%",
"(Reading database ... 65%",
"(Reading database ... 70%",
"(Reading database ... 75%",
"(Reading database ... 80%",
"(Reading database ... 85%",
"(Reading database ... 90%",
"(Reading database ... 95%",
"(Reading database ... 100%",
"(Reading database ... 143504 files and directories currently installed.)",
"Preparing to unpack .../gitlab-ce_13.2.9-ce.0_amd64.deb ...",
"/var/lib/dpkg/tmp.ci/preinst: 41: /var/lib/dpkg/tmp.ci/preinst: gitlab-ctl: not found",
"dpkg: error processing archive /var/cache/apt/archives/gitlab-ce_13.2.9-ce.0_amd64.deb (--unpack):",
" new gitlab-ce package pre-installation script subprocess returned error exit status 1",
"/var/lib/dpkg/info/gitlab-ce.postinst: 114: /var/lib/dpkg/info/gitlab-ce.postinst: /opt/gitlab/embedded/bin/symlink_ctl_cmds: not found",
"cp: cannot stat '/opt/gitlab/etc/gitlab.rb.template': No such file or directory",
"sed: can't read /etc/gitlab/gitlab.rb: No such file or directory",
"chmod: cannot access '/etc/gitlab/gitlab.rb': No such file or directory",
"Errors were encountered while processing:",
" /var/cache/apt/archives/gitlab-ce_13.2.9-ce.0_amd64.deb",
"needrestart is being skipped since dpkg has failed"]}
```
   
In summary, it fails because the `/usr/bin/gitlab-ctl` file is not found and the instalation of the package fail.
We have tried manualy and it's the same thing.  
  
We propose an improvement:  

It would be necessary to create a task which compares the installed version and the version indicated in the variable `gitlab_version`. if the version are diferent execute the task  [`Install GitLab`](https://github.com/geerlingguy/ansible-role-gitlab/blob/374dcf1d2a57d64ee83b9733af532ba51e45445c/tasks/main.yml#L42) 

Whe have do a aditionnal task on our playbook.

We ca propose a pull request if you validate this improvement.

Thank's the Lydra team

**Comment by t-readyroc** [comment]
As I'm about to make the same change to add the necessary tasks, I think this is very much a worthwhile addition.

**Comment by drschipm** [comment]
I also agree this would be very valuable. I am being tasked to utilize Ansible for future Gitlab upgrades (Omnibus install). The playbook works great for initial installs but I haven't utilized it for upgrades. Glad to see there is already a discussion about this.

Thanks again for all the work on this @geerlingguy.

**Comment by geerlingguy** [comment]
If it would be a simple enough addition, I'd be happy to include it; would make it even easier to manage GitLab instances.

**Comment by geerlingguy** [comment]
I should note it looks like there's a PR for this already here https://github.com/geerlingguy/ansible-role-gitlab/pull/161

**Comment by cchaudier** [comment]
Hi, I can propose you a PR.
We have do this with post_tasks on our playbook for the moment.

**Comment by t-readyroc** [comment]
@geerlingguy looks like the PR you mentioned is closed due to inactivity.

**Comment by geerlingguy** [comment]
@t-readyroc - heh, just two days after I commented on it too... I reopened it for now.

**Comment by drschipm** [comment]
@cchaudier could you elaborate on what type of post_tasks you have setup to accomplish this. I'm also investigating PR #161 but was curious what kind of post tasks were involved in your solution. Thanks

**Comment by cchaudier** [comment]
Hello @drschipm,

For the moment we have di this :
```
  roles:
    - geerlingguy.gitlab

  post_tasks:
    - name: "Check if the installed version of gitlab is {{ gitlab_version_number }}"
      lineinfile:
        path:   /opt/gitlab/version-manifest.txt
        regexp: "{{ gitlab_edition }}.*{{ gitlab_version_number }}"
        state: absent
      check_mode: true
      changed_when: false
      register: same_gitlab_version

    - import_tasks: tasks/gitlab_upgrade.yml
      when: not same_gitlab_version.found

    - name: Create mattermost configuration file
      template:
        src: "{{ gitlab_mattermost_config_template }}.j2"
        dest: "{{ gitlab_mattermost_config_path }}"
        owner: mattermost
        group: root
        mode: 0600
      notify: restart gitlab
```

```tasks/gitlab_upgrade.yml
# tasks/gitlab_upgrade.yml

- name: "Update: Define the Gitlab package name"
  set_fact:
    gitlab_package_name: "{{ gitlab_edition }}{{ gitlab_package_version_separator }}{{ gitlab_version }}"

- name: "Update GitLab"
  package:
    name: "{{ gitlab_package_name | default(gitlab_edition) }}"
    state: present
  notify: restart gitlab
```

**Comment by drschipm** [comment]
@cchaudier Thanks for the update! I'll work on adding this to my playbook

**Comment by t-readyroc** [comment]
@geerlingguy It doesn't look like #161 was ever merged?

**Comment by djetelina** [comment]
Seems like this was never solved :(
